device/iot-device-samples/multiplexing-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/ClientManagerBase.java

```diff
@@ -0,0 +1,166 @@
+package samples.com.microsoft.azure.sdk.iot;
+
+import com.microsoft.azure.sdk.iot.device.DeviceTwin.Pair;
+import com.microsoft.azure.sdk.iot.device.IotHubConnectionStatusChangeCallback;
+import com.microsoft.azure.sdk.iot.device.IotHubConnectionStatusChangeReason;
+import com.microsoft.azure.sdk.iot.device.exceptions.DeviceOperationTimeoutException;
+import com.microsoft.azure.sdk.iot.device.exceptions.TransportException;
+import com.microsoft.azure.sdk.iot.device.transport.IotHubConnectionStatus;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.IOException;
+
+import static com.microsoft.azure.sdk.iot.device.IotHubConnectionStatusChangeReason.NO_NETWORK;
+import static com.microsoft.azure.sdk.iot.device.IotHubConnectionStatusChangeReason.RETRY_EXPIRED;
+import static com.microsoft.azure.sdk.iot.device.transport.IotHubConnectionStatus.DISCONNECTED;
+import static com.microsoft.azure.sdk.iot.device.transport.IotHubConnectionStatus.DISCONNECTED_RETRYING;
+
+/**
+ * This is the base class for DeviceClientManager and MultiplexingClientManager
+ * The shared logic for reconnection and handling status change call back exists here.
+ */
+@Slf4j
+public abstract class ClientManagerBase implements IotHubConnectionStatusChangeCallback {
+
+    protected enum ConnectionStatus {
+        DISCONNECTED, CONNECTING, CONNECTED
+    }
+    
+    protected static final Object lock = new Object();
+    protected static final int SLEEP_TIME_BEFORE_RECONNECTING_IN_SECONDS = 2;
+    protected ConnectionStatus connectionStatus;
+    protected Pair<IotHubConnectionStatusChangeCallback, Object> suppliedConnectionStatusChangeCallback;
+
+    public abstract void openClient() throws IOException;
+    public abstract void closeClient() throws IOException;
+    public abstract String getClientId();
+
+    public void registerConnectionStatusChangeCallback(IotHubConnectionStatusChangeCallback callback, Object callbackContext) {
+        if (callback != null) {
+            this.suppliedConnectionStatusChangeCallback = new Pair<>(callback, callbackContext);
+        } else {
+            this.suppliedConnectionStatusChangeCallback = null;
+        }
+    }
+
+    public void doConnect() throws IOException {
+        // Device client does not have retry on the initial open() call. Will need to be re-opened by the calling application
+        while (connectionStatus == ConnectionStatus.CONNECTING) {
+            synchronized (lock) {
+                if(connectionStatus == ConnectionStatus.CONNECTING) {
```

Comment:
Suggest changing this to != connecting, and return, so the remaining code doesn't have to be tabbed in 1 more